src/binarySearchTree/bst.js

```diff
@@ -0,0 +1,516 @@
+      this.data.forEach(function (d) {
+        if (!self.checkValueEquality(d, value)) { newData.push(d) }
+      })
+      self.data = newData
+      return
+    }
+
+    // Delete the whole node
+    if (this.deleteIfLeaf()) {
+      return
+    }
+    if (this.deleteIfOnlyOneChild()) {
+      return
+    }
+
+    // We are in the case where the node to delete has two children
+    if (Math.random() >= 0.5) {   // Randomize replacement to avoid unbalancing the tree too much
+                                  // Use the in-order predecessor
+      replaceWith = this.left.getMaxKeyDescendant()
+
+      this.key = replaceWith.key
+      this.data = replaceWith.data
+
+      if (this === replaceWith.parent) {   // Special case
+        this.left = replaceWith.left
+        if (replaceWith.left) { replaceWith.left.parent = replaceWith.parent }
+      } else {
+        replaceWith.parent.right = replaceWith.left
+        if (replaceWith.left) { replaceWith.left.parent = replaceWith.parent }
+      }
+    } else {
+      // Use the in-order successor
+      replaceWith = this.right.getMinKeyDescendant()
+
+      this.key = replaceWith.key
+      this.data = replaceWith.data
+
+      if (this === replaceWith.parent) {   // Special case
+        this.right = replaceWith.right
+        if (replaceWith.right) { replaceWith.right.parent = replaceWith.parent }
+      } else {
+        replaceWith.parent.left = replaceWith.right
+        if (replaceWith.right) { replaceWith.right.parent = replaceWith.parent }
+      }
+    }
+  }
+
+  /**
+   * Execute a function on every node of the tree, in key order
+   * @param {Function} fn Signature: node. Most useful will probably be node.key and node.data
+   */
+  executeOnEveryNode(fn) {
+    if (this.left) { this.left.executeOnEveryNode(fn) }
+    fn(this)
+    if (this.right) { this.right.executeOnEveryNode(fn) }
+  }
+
+  /**
+   * Pretty print a tree
+   * @param {Boolean} printData To print the nodes' data along with the key
+   */
+  prettyPrint(printData, spacing) {
+    spacing = spacing || ''
+
+    console.log(spacing + '* ' + this.key)
+    if (printData) { console.log(spacing + '* ' + this.data) }
+
+    if (!this.left && !this.right) { return }
+
+    if (this.left) {
+      this.left.prettyPrint(printData, spacing + '  ')
+    } else {
+      console.log(spacing + '  *')
+    }
+    if (this.right) {
+      this.right.prettyPrint(printData, spacing + '  ')
+    } else {
+      console.log(spacing + '  *')
+    }
+  }
+}
+
+// ================================
+// Methods used to test the tree
+// ================================
+
+// ============================================
+// Methods used to actually work on the tree
+// ============================================
+
+// Append all elements in toAppend to array
+function append (array, toAppend) {
+  let i
+
+  for (i = 0; i < toAppend.length; i += 1) {
+    array.push(toAppend[i])
+  }
+}
```

Comment:
well this function just makes no sense in modern JS, could be replaced with `Array.prototype.concat`